File: services/collectors/rivalskins/feed_fetcher.py
"""Fetch and parse the rivalskins.com leaks RSS feed (a WordPress RSS 2.0 feed).

rivalskins.com posts full-resolution skin renders a few days before a patch. The
feed needs no special User-Agent. Each <item> carries the skin name, the post URL,
a pubDate and the body in <content:encoded>; the first non-banner image in that
HTML is the skin render.
"""

from __future__ import annotations

import logging
import re
from dataclasses import dataclass
from datetime import timezone
from email.utils import parsedate_to_datetime
from urllib.parse import urlsplit
from xml.etree import ElementTree

import httpx
from bs4 import BeautifulSoup


logger = logging.getLogger(__name__)

DEFAULT_FEED_URL = "https://rivalskins.com/category/leaks/feed/"
USER_AGENT = "MarvelRivalsUACollector/1.0 (Telegram news bot; +https://t.me/MarvelRivalsUABot)"
REQUEST_TIMEOUT_SECONDS = 25.0

_CONTENT = "{http://purl.org/rss/1.0/modules/content/}"

_IMAGE_EXT_RE = re.compile(r"\.(?:jpe?g|png|webp)(?:\?|$)", re.IGNORECASE)
# The site's own host (incl. any subdomain/CDN). Only images here are trusted as
# the post's photo, so a third-party <img> embedded in the body is never sent.
_IMAGE_HOST_SUFFIX = "rivalskins.com"
# A recurring season-launch banner that appears in every post — never the unique
# skin render, so it is skipped when picking the post image.
_BANNER_RE = re.compile(r"launch-skins", re.IGNORECASE)


@dataclass(frozen=True)
class RivalSkinsPost:
    post_id: str  # the feed guid (?p=<id>), stable across edits
    web_url: str
    title: str
    body_text: str
    created_at: str | None
    image_url: str | None


class RivalSkinsFeedFetcher:
    def __init__(self, feed_url: str) -> None:
        self.feed_url = feed_url

    async def fetch_recent_posts(self) -> list[RivalSkinsPost]:
        xml_text = await self._fetch_feed()
        if xml_text is None:
            return []

        posts = parse_feed(xml_text)
        if not posts:
            logger.warning("rivalskins feed %s returned no usable posts", self.feed_url)
        else:
            logger.info("Fetched %s rivalskins posts", len(posts))
        return posts

    async def _fetch_feed(self) -> str | None:
        try:
            async with httpx.AsyncClient(
                follow_redirects=True,
                timeout=httpx.Timeout(REQUEST_TIMEOUT_SECONDS),
                headers={"User-Agent": USER_AGENT, "Accept": "application/rss+xml, application/xml"},
            ) as client:
                response = await client.get(self.feed_url)
                response.raise_for_status()
                return response.text
        except httpx.HTTPError as exc:
            logger.warning("Failed to fetch rivalskins feed %s: %s", self.feed_url, exc)
            return None


def parse_feed(xml_text: str) -> list[RivalSkinsPost]:
    try:
        root = ElementTree.fromstring(xml_text)
    except ElementTree.ParseError as exc:
        logger.warning("Failed to parse rivalskins feed XML: %s", exc)
        return []

    channel = root.find("channel")
    if channel is None:
        return []

    posts: list[RivalSkinsPost] = []
    seen_ids: set[str] = set()
    for item in channel.findall("item"):
        post = _parse_item(item)
        if post is None or post.post_id in seen_ids:
            continue
        seen_ids.add(post.post_id)
        posts.append(post)

    return posts


def _parse_item(item: ElementTree.Element) -> RivalSkinsPost | None:
    link = (item.findtext("link") or "").strip()
    guid = (item.findtext("guid") or "").strip()
    post_id = guid or link
    if not post_id or not link:
        return None

    title = (item.findtext("title") or "").strip()
    created_at = _normalize_pubdate(item.findtext("pubDate"))
    body_text, image_url = _parse_content(item.findtext(f"{_CONTENT}encoded"))

    return RivalSkinsPost(
        post_id=post_id,
        web_url=link,
        title=title,
        body_text=body_text,
        created_at=created_at,
        image_url=image_url,
    )


def _parse_content(content_html: str | None) -> tuple[str, str | None]:
    """Return (clean body text, the skin-render image URL) from a post's HTML."""
    if not content_html or not content_html.strip():
        return "", None

    soup = BeautifulSoup(content_html, "html.parser")
    body_text = soup.get_text(" ", strip=True)

    image_url: str | None = None
    for anchor in soup.find_all("img", src=True):
        src = str(anchor["src"]).strip()
        if _is_allowed_image(src) and not _BANNER_RE.search(src):
            image_url = src
            break

    return body_text, image_url


def _is_allowed_image(url: str) -> bool:
    parsed = urlsplit(url.strip())
    host = parsed.netloc.lower()
    on_site = host == _IMAGE_HOST_SUFFIX or host.endswith("." + _IMAGE_HOST_SUFFIX)
    return parsed.scheme == "https" and on_site and bool(_IMAGE_EXT_RE.search(url))


def _normalize_pubdate(value: str | None) -> str | None:
    """RFC-822 pubDate -> a UTC ISO timestamp datetime.fromisoformat can parse."""
    if not value or not value.strip():
        return None
    try:
        parsed = parsedate_to_datetime(value.strip())
    except (TypeError, ValueError):
        return None
    if parsed is None:
        return None
    if parsed.tzinfo is None:
        parsed = parsed.replace(tzinfo=timezone.utc)
    return parsed.astimezone(timezone.utc).isoformat()

File: services/collectors/rivalskins/feed_fetcher.js
/**
 * Fetch and parse the rivalskins.com leaks RSS feed (a WordPress RSS 2.0 feed).
 *
 * rivalskins.com posts full-resolution skin renders a few days before a patch.
 * The feed needs no special User-Agent. Each <item> carries the skin name, the
 * post URL, a pubDate and the body in <content:encoded>; the non-banner images in
 * that HTML are the post's renders, and a post with two or more of them is
 * published as one album.
 */

import { DateTime } from "luxon";

import { getText, loadHtml } from "../../html.js";
import { getLogger } from "../../logger.js";
import { errorText } from "../../pyutils.js";
import { urlsplit } from "../../urlutils.js";
import { child, children, findText, parseXml, XmlParseError } from "../../xml.js";

const logger = getLogger("services.collectors.rivalskins.feed_fetcher");

export const DEFAULT_FEED_URL = "https://rivalskins.com/category/leaks/feed/";
export const USER_AGENT = "MarvelRivalsUACollector/1.0 (Telegram news bot; +https://t.me/MarvelRivalsUABot)";
export const REQUEST_TIMEOUT_SECONDS = 25.0;

const IMAGE_EXT_RE = /\.(?:jpe?g|png|webp)(?:\?|$)/i;
// The site's own host (incl. any subdomain/CDN). Only images here are trusted as
// the post's photo, so a third-party <img> embedded in the body is never sent.
const IMAGE_HOST_SUFFIX = "rivalskins.com";
// A recurring season-launch banner that appears in every post — never the unique
// skin render, so it is skipped when picking the post image.
const BANNER_RE = /launch-skins/i;
// Telegram's media-group maximum; a post never has this many renders, but the
// cap keeps a malformed body from queueing an unbounded album.
const MAX_IMAGES_PER_POST = 10;

/**
 * @typedef {{post_id: string, web_url: string, title: string, body_text: string,
 *            created_at: string|null, image_url: string|null,
 *            additional_image_urls: string[]}} RivalSkinsPost
 */

export class RivalSkinsFeedFetcher {
  constructor(feedUrl) {
    this.feed_url = feedUrl;
  }

  async fetchRecentPosts() {
    const xmlText = await this.fetchFeed();
    if (xmlText === null) {
      return [];
    }

    const posts = parseFeed(xmlText);
    if (!posts.length) {
      logger.warning(`rivalskins feed ${this.feed_url} returned no usable posts`);
    } else {
      logger.info(`Fetched ${posts.length} rivalskins posts`);
    }
    return posts;
  }

  async fetchFeed() {
    try {
      const response = await fetch(this.feed_url, {
        redirect: "follow",
        headers: { "User-Agent": USER_AGENT, Accept: "application/rss+xml, application/xml" },
        signal: AbortSignal.timeout(REQUEST_TIMEOUT_SECONDS * 1000),
      });
      if (!response.ok) {
        throw new Error(`HTTP ${response.status}`);
      }
      return await response.text();
    } catch (error) {
      logger.warning(`Failed to fetch rivalskins feed ${this.feed_url}: ${errorText(error)}`);
      return null;
    }
  }
}

export function parseFeed(xmlText) {
  let document;
  try {
    // The feed is untrusted, so DTDs/entity expansion (billion-laughs) and
    // external entities are refused up front.
    document = parseXml(xmlText);
  } catch (error) {
    if (!(error instanceof XmlParseError)) {
      throw error;
    }
    logger.warning(`Failed to parse rivalskins feed XML: ${errorText(error)}`);
    return [];
  }

  const root = child(document, "rss");
  const channel = root === null ? null : child(root, "channel");
  if (channel === null) {
    return [];
  }

  const parsed = [];
  const seenIds = new Set();
  for (const item of children(channel, "item")) {
    const post = parseItem(item);
    if (post === null || seenIds.has(post.post_id)) {
      continue;
    }
    seenIds.add(post.post_id);
    parsed.push(post);
  }

  return parsed.map((post) => finalizePost(post, recurringImages(parsed)));
}

/**
 * The image URLs that appear in more than one post of this fetch.
 *
 * Most posts end with the season-roadmap or Twitch-drop promo of the moment,
 * which is not that post's content: publishing it would put the same banner under
 * every skin leak. A render belongs to exactly one post, so showing up twice in
 * one feed is the signal — and it keeps working when the site swaps the banner
 * for a new one, which a hardcoded filename pattern would not.
 */
function recurringImages(posts) {
  const counts = new Map();
  for (const post of posts) {
    for (const url of new Set(post.images)) {
      counts.set(url, (counts.get(url) ?? 0) + 1);
    }
  }
  return new Set([...counts].filter(([, count]) => count > 1).map(([url]) => url));
}

function finalizePost({ images: allImages, ...fields }, recurring) {
  const images = allImages.filter((url) => !recurring.has(url)).slice(0, MAX_IMAGES_PER_POST);
  return Object.freeze({
    ...fields,
    image_url: images[0] ?? null,
    additional_image_urls: images.slice(1),
  });
}

function parseItem(item) {
  const link = findText(item, "link").trim();
  const guid = findText(item, "guid").trim();
  const postId = guid || link;
  if (!postId || !link) {
    return null;
  }

  const title = findText(item, "title").trim();
  const createdAt = normalizePubDate(findText(item, "pubDate"));
  const [bodyText, images] = parseContent(findText(item, "encoded"));

  return {
    post_id: postId,
    web_url: link,
    title,
    body_text: bodyText,
    created_at: createdAt,
    images,
  };
}

/** Return [clean body text, the post's on-site render URLs] from a post's HTML. */
export function parseContent(contentHtml) {
  if (!contentHtml || !contentHtml.trim()) {
    return ["", []];
  }

  const $ = loadHtml(contentHtml);
  const bodyText = getText($.root());

  const images = [];
  const seen = new Set();
  for (const image of $("img[src]").toArray()) {
    const src = String($(image).attr("src")).trim();
    if (!isAllowedImage(src) || BANNER_RE.test(src) || seen.has(src)) {
      continue;
    }
    seen.add(src);
    images.push(src);
  }

  return [bodyText, images];
}

function isAllowedImage(url) {
  const parsed = urlsplit(String(url).trim());
  const host = parsed.netloc.toLowerCase();
  const onSite = host === IMAGE_HOST_SUFFIX || host.endsWith(`.${IMAGE_HOST_SUFFIX}`);
  return parsed.scheme === "https" && onSite && IMAGE_EXT_RE.test(url);
}

/** RFC-822 pubDate -> a UTC ISO timestamp `datetime.fromisoformat` can parse. */
export function normalizePubDate(value) {
  if (!value || !value.trim()) {
    return null;
  }
  const parsed = DateTime.fromRFC2822(value.trim(), { setZone: true });
  if (!parsed.isValid) {
    return null;
  }
  // "+00:00", not luxon's default "Z": this string is stored in seen_sources and
  // compared against dates written by every other collector, all of which use
  // Python's `isoformat()` spelling of the UTC offset.
  return parsed.toUTC().toFormat("yyyy-MM-dd'T'HH:mm:ssZZ");
}
